File: revision/PredictTheWinner.java
public class PredictTheWinner {

    public static boolean predictTheWinner(int[] nums) {
        return maxDiff(0, nums.length - 1, nums) >= 0;
    }

    static int maxDiff(int left, int right, int[] nums) {
        if (left == right) {
            return nums[left];
        }
        int leftChoice = nums[left] - maxDiff(left + 1, right, nums);
        int rightChoice = nums[right] - maxDiff(left, right - 1, nums);
        return Math.max(leftChoice, rightChoice);
    }

    public static void main(String[] args) {
        System.out.println(predictTheWinner(new int[] { 1, 5, 2 }));
        System.out.println(predictTheWinner(new int[] { 1, 5, 233, 7 }));
    }
}

File: leetcode/predict-the-winner.py
class Solution:
    def predictTheWinner(self, nums: List[int]) -> bool:
        def diff(l, r, nums):
            if l == r:
                return nums[l]
            return max(nums[l]-diff(l+1, r, nums), nums[r]-diff(l, r-1, nums))
        return diff(0, len(nums)-1, nums) >= 0
